Integrate vtk9 support (phase1)

biagas opened the pull request:
### Description

This is the first in a series of PR's I will create that integrates the VTK-9 port into develop while maintaining support for VTK-8.
  
The desire is to produce a number of PR's of smaller size than the full port to VTK9..
This will make reviewing of PR's less onerous on reviewers, as it will reduce the number of files in a PR.

Smaller # of files/changes in each PR will also engender more discussion on choices made and could lead to improvements.

Maintaining support for VTK8 means that other development work won't be hampered by the changes necessary for VTK 9.

 ** VTK 9 support will not fully compile until all PR's for the integration have been reviewed and committed. **


<!-- Please include a summary of the change and any other information and instructions to help reviewers understand the work -->

This PR 
* adds Find module for VTK 9
* adds vtk9 support to build_visit.
* adds LIB_VERSION_LE, LIB_VERSION_GE macros to aid in conditional compilation based on VTK version
* modifies the calling order of Find modules for TP libs that VTK depends on, so that they are found and recorded before VTK is found (VTK 9 calls find for any TP libs it needs, and we want to be sure it finds the ones we specify).


### Type of change

<!-- Please check one of the boxes below -->

~~* [ ] Bug fix~~
* [X] New feature~~
~~* [ ] Documentation update~~
~~* [ ] Other~~ <!-- please explain with a note below -->

### How Has This Been Tested?

I have verified that build_visit and compilation of visit itself still work with VTK 8.

### Checklist:

<!-- For items in this checklist that do not apply, simply insert two tilde chars, `~~`, just ahead of the left bracket char, `[` at the beginning of a line. Each line ends with two tilde chars to make doing such ~~strikeouts~~ easy. -->

- [X] I have followed the [style guidelines][1] of this project.~~
- [X] I have performed a self-review of my own code.~~
- [X] I have commented my code where applicable.~~
~~- [ ] I have updated the release notes.~~
~~- [ ] I have made corresponding changes to the documentation.~~
~~- [ ] I have added debugging support to my changes.~~
~~- [ ] I have added tests that prove my fix is effective or that my feature works.~~
- [X] I have confirmed new and existing unit tests pass locally with my changes.~~
~~- [ ] I have added new baselines for any new tests to the repo.~~
~~- [ ] I have NOT made any changes to [*protocol* or *public interfaces*][3] in an RC branch.~~
- [X] I have assigned reviewers (see [VisIt's PR procedures][2] for more information).~~

[1]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/StyleGuide.html
[2]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/pr_create.html#reviewers
[3]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/RCDevelopment.html#communication-protocols-and-public-apis

biagas committed:
Add find module for VTK 9.

Update FindJPEG so it works with VTK 9.
Fix issue with install library macro where it wouldn't install libvtkXXX-9.1.0.so.1.
Remove VTK_INCLUDE_DIRS from plugins include list if NOT using vtk-8.

biagas committed:
Support VTK 9 in build_visit.

add --vtk9 option.
Modify bv_vtk to build vtk9 when requested.
Modify bv_xdmf to point to vtk9 when requested.

biagas committed:
Add LIB_VERSION_LE, LIB_VERSION_GE macros.

Use in conjunction with SETUP_APP_VERISON which creates the XXX_VERSION_HEX for a library.
The macros allow for conditional compilation based on library version.
Update Uintah to use new lib version macros which replaced APP_VERSION_CHECK.

biagas committed:
Re-order calling of Find modules in root CMakeLists.txt

VTK9's now calls its own find for dependent TP libs, so we need to make sure our
find modules for those libs are called before VTK's, so that our versions are found and used.
Update windows config-site for VTK9.

JustinPrivitera commented on src/CMake/VisItIncludeVars.cmake:
```diff
@@ -50,9 +50,12 @@ set(VISIT_PLUGIN_INCLUDES
-    ${VTK_INCLUDE_DIRS} 
 )
 
+if(VTK_VERSION VERSION_LESS_EQUAL "8.1.0")
```
I trust that this is right, but it doesn't make sense to me when looking at it. Earlier I believe you said in a different file that "Only VTK versions 8.1 and 9.1 are supported", but here it looks like you're checking that the VTK version is less than or equal to 8.1. Is that right?

biagas replied:
I can see your point. I suppose the version check could be VERSION_EQUAL instead.  Might be more important when testing for '9.1.0' because in our root CMakeLists.txt we specify the minimum VTK version supported is 8.1.0. So any version < 8.1.0 will fail first there.
I will change the VERSION_GREATER_EQUAL '9.1.0' test I have, because I don't know if 9.1.x or 9.2 will work.

biagas replied:
The next PR will have changes to a lot of CMakeLists.txt with a version check.
Which do you think is better:

```
if(VTK_VERSION VERSION_EQUAL "8.1.0")
  do_something_for_vtk_8()
else()
  do_something_for_vtk9()
endif()
```
or

```
if(VTK_VERSION VERSION_GREATER_EQUAL "9.1.0")
    do_something_for_vtk9()
else()
    do_something_for_vtk8()
endif()
```

brugger1 replied:
I prefer the first, since the normal case right now is vtk8.

JustinPrivitera commented on src/CMakeLists.txt:
```diff
@@ -1276,86 +1304,75 @@ ELSE(VISIT_THREAD)
 
 # Configure PySide Support
-if(VISIT_PYTHON_SCRIPTING AND PYTHONLIBS_FOUND
+if(VTK_VERSION VERSION_LESS_EQUAL "8.1.0")
```
same question here as before

JustinPrivitera commented on src/tools/dev/scripts/bv_support/bv_main.sh:
```diff
@@ -1238,7 +1242,9 @@ function run_build_visit()
-            --console) ;;
+            # "--vtk9" is actually handled elsewhere, but it is also here
+            # to prevent it triggering an "Urecognized option" error.
+            --vtk9) ;;
```
Do we document our `build_visit` options anywhere? If we do I can't find it. I think it'd be good if we did.

biagas replied:
build_visit --help should show all the options. I just checked and my addition of --vtk9 isn't listed, so I will correct that oversight.

biagas replied:
Fixed in bv_vtk.sh to show '--vtk9' as another vtk option to be used in conjunction with --vtk.

JustinPrivitera commented on src/tools/dev/scripts/bv_support/bv_vtk.sh:
```diff
@@ -1432,52 +1954,94 @@ EOF
-        return 1
-    fi
+    if [[ "$DO_VTK9" == "yes" ]] ; then
+        # vtk8 version needs to be reworked for 9.1.0
+        #apply_vtk9_vtkopenfoamreader_patch
+        #if [[ $? != 0 ]] ; then
+        #    return 1
+        #fi
+
+        # vtk8 version needs to be reworked for 9.1.0
+        #apply_vtk9_vtkopenglspheremapper_h_patch
+        #if [[ $? != 0 ]] ; then
+        #    return 1
+        #fi
+
+        # vtk8 version needs to be reworked for 9.1.0
+        #apply_vtk9_vtkopenglspheremapper_patch
+        #if [[ $? != 0 ]] ; then
+        #    return 1
+        #fi
```
Are these supposed to still be commented? Are they for the next PR in the series?

biagas replied:
Yes, they are supposed to still be commented out. I haven't worked out if the patches are still needed and what form they would take in VTK9.

biagas committed:
Add --vtk9 to bv_vtk_print_usage

biagas committed:
Change VERSION_LESS_EQUAL test to VERSION_EQUAL

biagas committed:
Change VERSION_LESS_EQUAL test to VERSION_EQUAL.

biagas committed:
Add specific version test to FindVisItVTK.cmake

Only 8.1 in the 8.x series is supported, and only 9.1 in the 9.x series, so make this explicit until newer versions of 9.x are supported.

biagas committed:
Merge branch 'develop' into task/biagas/add_vtk9_support_phase1

biagas committed:
Merge branch 'develop' into task/biagas/add_vtk9_support_phase1
